db/migration/migration.go

```diff
@@ -126,25 +120,51 @@ func Run(subCmd string, args []string) error {
 	dbType, dbConnection, migrationsPath, _ := readGooseConfig()
 
 	if err := goose.SetDialect(dbType); err != nil {
-		fmt.Printf("error: failed to set goose dialect: %s\n", err)
-		return err
+		return fmt.Errorf("failed to set goose dialect: %s", err)
 	}
 
 	db, err := sql.Open(dbType, dbConnection)
 
 	if err != nil {
-		fmt.Printf("error: failed to open db: %s\n", err)
-		return err
+		return fmt.Errorf("failed to open db: %s", err)
 	}
 
 	if err = goose.LoadMigrationPlugins(migrationsPath); err != nil {
-		logger.Error("migration: failed to load migration plugins: %s", err)
-		return err
+		return fmt.Errorf("failed to load migration plugins: %s", err)
+	}
+
+	if err := lock(dbType, db); err != nil {
+		return fmt.Errorf("failed to acquire migration lock: %s", err)
+	}
+	defer unlock(dbType, db)
+
+	if err := goose.Run(subCmd, db, migrationsPath, args...); err != nil {
+		return fmt.Errorf("migration failed to run: %s", err)
 	}
 
-	if err = goose.Run(subCmd, db, migrationsPath, args...); err != nil {
-		fmt.Printf("migration: failed to run: %s\n", err)
-		return err
+	return nil
+}
+
+// lock issues a lock on a shared mutex to avoid running migrations on the same
+// database in parallel, currently only supported on mysql.
+func lock(dbType string, db *sql.DB) error {
+	switch dbType {
+	case "mysql":
+		if _, err := db.Exec("LOCK TABLES `goose_db_version` WRITE"); err != nil {
```

Comment:
this may not be the right thing to do, I think we have to use etcd instead:
> LOCK TABLES doesn't work when using Galera cluster. You may experience crashes or locks when used with Galera.

https://mariadb.com/kb/en/mariadb/lock-tables-and-unlock-tables/